src/Http/Http.Features/src/IHttpRequestBodyDetectionFeature.cs

```diff
@@ -0,0 +1,27 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Http.Features
+{
+    /// <summary>
+    /// Used to indicate if the request can have a body.
+    /// </summary>
+    public interface IHttpRequestBodyDetectionFeature
+    {
+        /// <summary>
+        /// Indicates if the request can have a body.
+        /// </summary>
+        /// <remarks>
+        /// This returns true when:
+        /// - It's an HTTP/1.x request with a non-zero Content-Length or a 'Transfer-Encoding: chunked' header.
+        /// - It's an HTTP/2 request that did not set the END_STREAM flag on the initial headers frame.
+        /// The final request body length may still be zero for the chunked or HTTP/2 scenarios.
+        ///
+        /// This returns false when:
+        /// - It's an HTTP/1.x request with no Content-Length or 'Transfer-Encoding: chunked' header, or the Content-Length is 0.
+        /// - It's an HTTP/2 request that set END_STREAM on the initial headers frame.
+        /// When false, the request body should never return data.
+        /// </remarks>
```

Comment:
As written, this will also return false for `Connection: Upgrade` requests in Kestrel. See https://github.com/dotnet/aspnetcore/blob/4ecde5f6b0d1aba59c7bd0646e20e7b16a78f767/src/Servers/Kestrel/Core/src/Internal/Http/Http1UpgradeMessageBody.cs#L23-L24

This is despite the fact that the request body can have content and is very much readable and was only written that way to avoid unnecessary drain timeouts since previously MessageBody.IsEmpty was only used for Kestrel to determine whether it should skip trying to drain the request body.

We should probably have separate `ShouldDrain` and `IsEmpty` properties on MessageBody and add a test for this scenario.

Reply:
False is correct for Connection: Upgrade requests, there's no normal HTTP request body, and there shouldn't be any data on the connection until after the upgrade. I can call that out in the comments.

It's still pretty weird that Kestrel allows you to read from the regular body stream after the upgrade vs the stream returned from upgrade, I don't think the other servers do.